### PR TITLE
[repo] Mitigate vulnerabilities in System.Text.Json 8.0.0 packages

### PR DESCRIPTION
**Pull request by CodeBlanch**
## Changes

* Mitigate vulnerabilities in System.Text.Json v8.0.0 - 8.0.3

## Details

ConsoleExporter and ZipkinExporter use System.Text.Json (STJ) but don't have a reference to it for `net8.0`+ targets. What happens is they get STJ transitively via `Microsoft.NETCore.App` framework reference. The final version will depend on the runtime version deployed with the app.

The problem is STJ v8.0.0 - 8.0.3 have been deprecated due to a [deserialization vulnerability](https://github.com/advisories/GHSA-hh2w-p6rv-4g7w).

The goal here is to redirect STJ to v8.0.4 for `net8.0` targets. Older targets should stay on v4.7.2. Newer targets (`net9.0`) will continue to use the transitive reference.

#### 1.9.0 stable:

|Target|Direct reference(s)|Framework reference|Version|Vulnerable|Notes|
|---|---|---|---|---|---|
|net462|System.Text.Encodings.Web & System.Text.Json||v4.7.2|No||
|netstandard2.0|System.Text.Encodings.Web & System.Text.Json||v4.7.2|No||
|net6.0||System.Text.Json|Runtime version (6.0.0 - 6.0.9)|No|Version depends on patch level of runtime|
|net8.0||System.Text.Json|Runtime version (8.0.0 - 8.0.4)|When <= 8.0.3|Version depends on patch level of runtime|

#### 1.10.0-beta.1:
|Target|Direct reference(s)|Framework reference|Version|Vulnerable|Notes|
|---|---|---|---|---|---|
|net462|System.Text.Encodings.Web & System.Text.Json||v4.7.2|No||
|netstandard2.0|System.Text.Encodings.Web & System.Text.Json||v4.7.2|No||
|net8.0||System.Text.Json|Runtime version (8.0.0 - 8.0.4)|When <= 8.0.3|Version depends on patch level of runtime|
|net9.0||System.Text.Json|Runtime version (9.0.0)|No|No patches yet for .NET 9|

#### 1.9.0 stable:

No hot patch currently planned. The vulnerability is about deserialization of untrusted input which neither ConsoleExporter nor ZipkinExporter is susceptible to. I'm approaching this as a low severity issue but some work needs to be done to agree on a severity and publish an advisory. If we determine there is a higher severity we will do a hot patch for 1.9.0, possibly other releases.

#### Next release of 1.10.0:
|Target|Direct reference(s)|Framework reference|Version|Vulnerable|Notes|
|---|---|---|---|---|---|
|net462|System.Text.Encodings.Web & System.Text.Json||v4.7.2|No||
|netstandard2.0|System.Text.Encodings.Web & System.Text.Json||v4.7.2|No||
|net8.0|System.Text.Json||v8.0.4|No||
|net9.0||System.Text.Json|Runtime version (9.0.0)|No|No patches yet for .NET 9|

## Merge requirement checklist

* [X] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [X] Appropriate `CHANGELOG.md` files updated for non-trivial changes

### CODE REVIEW

File: Directory.Packages.props
Comment (Kielek) [inline Directory.Packages.props:60]: @CodeBlanch, if you are keeping it legacy only for AutoInstrumentation purposes - it is not the case. The bigger problem we have with upgrading packages for the .NET6+.

From this perspective I would keep it consistent and upgrade to 8.0.0/8.0.4 respectively both for .NET Framework and .NET. Then avoid any other upgrades until we really need this. So still be convervative.

If possible, it will be great to release together with other dependency-breaking changes like System.Diagnostics.DiagnosticSource 9 and avoiding patch release with only this bump.

**Review by reyang — APPROVED** [review]

File: Directory.Packages.props
Comment (reyang) [reply Directory.Packages.props:60]: > If possible, it will be great to release together with other dependency-breaking changes like System.Diagnostics.DiagnosticSource 9 and avoiding patch release with only this bump.

I'm a bit confused here, I think we've been clear about `System.Diagnostics.DiagnosticSource` - there is *no* breaking change during major version bumps.

Comment (Kielek) [reply Directory.Packages.props:60]: It is from the auto instrumentation users. Any common package changes can be problematic.

Comment (reyang) [reply Directory.Packages.props:60]: I see, that is a technical issue in the current auto-instrumentation design, we shouldn't mix it up and get ourselves confused by saying "dependency-breaking changes like System.Diagnostics.DiagnosticSource 9".

Comment (CodeBlanch) [reply Directory.Packages.props]: I tweaked it so our minimum version for System.Text.Json is v8.0.4. I decided to go that direction because there were no objections from the auto-instrumentation folks and doing so allowed us to also drop explicit references to System.Text.Encodings.Web which was a nice/welcome side-effect.

File: Directory.Packages.props
Comment (TimothyMothra) [inline Directory.Packages.props:6]: nit: Are you planning on renaming "OutOfBandVer" when .NET9 is GA?

I fully support parameterizing this value for ease of maintenance, just curious about the name.

File: src/OpenTelemetry.Exporter.Console/OpenTelemetry.Exporter.Console.csproj
Comment (TimothyMothra) [inline src/OpenTelemetry.Exporter.Console/OpenTelemetry.Exporter.Console.csproj:8]: nit: This seems funky to me and may be a slippery slope.

If I didn't know anything about this repo, I would first look at the PackageReferences to tell me about project dependencies and may not catch the significance of this item. I personally prefer the ItemGroup with Condition instead of the extra Targets file.

File: Directory.Packages.props
Comment (CodeBlanch) [reply Directory.Packages.props:6]: I used "OutOfBand" here because that's what runtime calls its libraries which are pushed to NuGet: https://github.com/dotnet/runtime/blob/main/src/libraries/README.md#deployment

Happy to rename it if you think something else would be better though.

**Review by reyang — DISMISSED** [review]
Several changes were made after my initial approval. The PR scope is now much bigger, I think we should focus on the mitigation and put everything else in other PRs.

File: src/OpenTelemetry.Exporter.Console/OpenTelemetry.Exporter.Console.csproj
Comment (TimothyMothra) [reply src/OpenTelemetry.Exporter.Console/OpenTelemetry.Exporter.Console.csproj:8]: Only a nit pick. I don't feel strongly enough to block the PR.
I understand why it was done and I don't disagree with the intention to keep things simple.
For me, I'm thinking about times I've needed to troubleshoot csproj files and something like this likely would have tripped me up.

**Comment by CodeBlanch** [comment]
@reyang I reverted some of the cleanup/re-org being done in `Directory.Packages.props` does that reduce the scope enough or you want more reduction?

File: Directory.Packages.props
Comment (reyang) [reply Directory.Packages.props]: > I tweaked it so our minimum version for System.Text.Json is v8.0.4. I decided to go that direction because there were no objections from the auto-instrumentation folks and doing so allowed us to also drop explicit references to System.Text.Encodings.Web which was a nice/welcome side-effect.

This is a red flag to me - the fact that auto-instrumentation folks are fine and it will be easier for OTel .NET maintainers cannot justify such change. @CodeBlanch could you evaluate the impact and share the results here?

File: Directory.Packages.props
Comment (CodeBlanch) [reply Directory.Packages.props]: Ended up removing this from the PR. I'll bring it back as a follow-up.

File: Directory.Packages.props
Comment (CodeBlanch) [reply Directory.Packages.props]: I've started a discussion with the STJ team to understand the impact of upgrading vs the risk of staying on v4.7.2 (which is no longer being maintained). Waiting to hear back.

In the meantime I've reverted the version bump where it isn't strictly necessary. Also updated the description with more details.

File: Directory.Packages.props
Comment (reyang) [inline Directory.Packages.props:64]: FYI @CodeBlanch if you need a better link, this might be one https://devblogs.microsoft.com/dotnet/system-text-json-in-dotnet-7/#breaking-changes

Comment (CodeBlanch) [reply Directory.Packages.props:64]: Updated thanks!

**Review by reyang — APPROVED** [review]

**Pull request merged**
